HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,440 @@
+      continue;
+    // apply all the cuts
+    if (!applySelection(L3cand, beamSpot, cutParsL3))
+      continue;
+    nL3++;
+    filterproduct.addObject(trigger::TriggerMuon, L3cand);
+  }
+  // get the L3 muons that passed the selection
+  std::vector<reco::RecoChargedCandidateRef> passingL3mucands;
+  filterproduct.getObjects(trigger::TriggerMuon, passingL3mucands);
+
+  // Number of objects passing the L2 Trigger:
+  int nL2 = 0;
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
+    reco::RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L2cand isNonnull " << L2cand.isNonnull();
+    // safeguard
+    if (L2cand->pt() == 0.0) {
+      LogDebug("HLTMuonL3orL2PreFilter") << "Skipping candidate: L2cand pT is 0";
+      continue;
+    }
+    // did this candidate triggered at previous stage.
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L2cand, vcands))
+      continue;
+    // check if L2 and L3 candidates are the same
+    if (checkOverlap(L2cand, passingL3mucands, maxDeltaR2_L2L3_)) {
+      continue;
+    }
+    // apply all the cuts
+    if (!applySelection(L2cand, beamSpot, cutParsL2))
+      continue;
+    nL2++;
+    filterproduct.addObject(trigger::TriggerMuon, L2cand);
+  }
+
+  // filter decision
+  const bool acceptL3(nL3 >= min_N_L3_);
+  const bool acceptL2(nL2 >= min_N_L2_);
+
+  LogDebug("HLTMuonL3orL2PreFilter") << " >>>>> Result of HLTMuonL3orL2PreFilter is " << (acceptL3 && acceptL2)
+                                     << ", number of muons passing thresholds= " << nL2 + nL3;
+
+  return acceptL2 && acceptL3;
+}
+
+bool HLTMuonL3orL2PreFilter::triggeredByPreviousLevel(const reco::RecoChargedCandidateRef& candref,
+                                                      const std::vector<reco::RecoChargedCandidateRef>& vcands) const {
+  return (std::find(vcands.begin(), vcands.end(), candref) != vcands.end());
+}
+
+bool HLTMuonL3orL2PreFilter::checkOverlap(
+    const reco::RecoChargedCandidateRef& r1,
+    const std::vector<reco::RecoChargedCandidateRef>& R2,  // vector with the recoCharged candidates from L3
+    const double& maxDeltaR2) const {
+  for (unsigned int iMu = 0; iMu < R2.size(); iMu++) {
+    reco::RecoChargedCandidateRef r2 = R2[iMu];
+    if (reco::deltaR2(*r1, *r2) < maxDeltaR2)
+      return true;
+  }
+  return false;
+}
+
+double HLTMuonL3orL2PreFilter::pTeffTransformation(const reco::RecoChargedCandidateRef& candidate,
+                                                   const reco::TrackRef& track,
+                                                   const cutParameters& cutPars) const {
+  auto const pt = candidate->pt();
+  auto const err0 = track->error(0);
+  auto const abspar0 = std::abs(track->parameter(0));
+  // if abspar0 > 0, convert 50% efficiency threshold to 90% efficiency threshold
+  auto const ptLx = abspar0 > 0 ? pt + cutPars.nsigma_Pt * err0 / abspar0 * pt : pt;
+  return ptLx;
+}
```

Comment:
```suggestion
```
In the latest version this is only called in one place, so it's easier to just move it inside `applySelection`.